### PR TITLE
*: reduce ResetContextOfStmt() and NewRuntimeStatsColl() object allocation

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Reduce prepared point get object allocation, see https://github.com/pingcap/tidb/issues/25573

Set `EnableCollectExecutionInfo` config to false can reduce the allocation from 101 allocs/op to 89 allocs/op, it does have some performance impact on the point get case @crazycs520 

### What is changed and how it works?

Reduce the allocation rate by 2%

Before:

BenchmarkPreparedPointGet-16    	 3945436	      9160 ns/op	    7830 B/op	     101 allocs/op

After:

BenchmarkPreparedPointGet-16    	 4047969	      8913 ns/op	    7038 B/op	      99 allocs/op

What's Changed:

Avoid allocating StatementContext every time, reuse it.
Avoid allocating RuntimeStatsColl every time, reuse it.

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- No code

### Release note 

- No release note

### CODE REVIEW

**Comment by tiancaiamao** [comment]
image

Each time we eliminate a allocation from the list, we can reduce the allocation by 2%-1%, and the performance improve a bit.
As estimated in https://github.com/pingcap/tidb/issues/25573#issuecomment-874429222, cut down about a half of the allocation can speed up the point get function by 35%

File: executor/executor.go [hunk 1]
```diff
@@ -1656,14 +1656,31 @@ func (e *UnionExec) Close() error {
...
+	if vars.TxnCtx.CouldRetry {
+		// Must construct new statement context object, the retry history need context for every statement.
+		// TODO: Maybe one day we can get rid of transaction retry, then this logic can be deleted.
+		vars.StmtCtx = &stmtctx.StatementContext{}
+	} else {
+		// Reuse mem to reduce object allocation.
+		vars.InitStatementContext()
 	}
+
+	sc := vars.StmtCtx
```
Comment (xhebox) [inline executor/executor.go:1676]: Will `sc = stmtctx.StatementContext{}` do the trick, I mean, reset the sc? I don't understand why tests need to be modified if logic is same. One thing I noticed is that, `&xxx{}` will init all fields into the default zero value. But the new logic only init fields that is manually inited, other fields are not reseted.

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: executor/executor.go (hunk 1 above)
Comment (tangenta) [reply executor/executor.go:1676]: > I don't understand why tests need to be modified

+1.

Comment (tiancaiamao) [reply executor/executor.go:1676]: > Will sc = stmtctx.StatementContext{} do the trick

Yes, we can. 
But there is a drawback, when the struct are nested, the allocation still can't be avoided.

```
type SessionVars struct {
    StmtCtx *stmtctx.StatementContext
}

type StatementContext struct {
    RuntimeStatsColl *execdetails.RuntimeStatsColl
}

*vars.StmtCtx = stmtctx.StatementContext{}  // reset the data and reuse memory

vars.StmtCtx.RuntimeStatsColl = NewRuntimeStatsColl() // after reset, it's gone ... can't avoid allocating a new one!
```

The method adopted in this PR can avoid allocation totally.

```
type SessionVars struct {
    StmtCtx *stmtctx.StatementContext // point to the cached data
    cache struct {
        stmtctx.StatementContext
    }
}

type StatementContext struct {
    RuntimeStatsColl *execdetails.RuntimeStatsColl  // point to the cached data
    cache struct {
        RuntimeStatsColl
    }
}

vars.cache.StatementContext = stmtctx.StatementContext{}  // reset data
vars.StmtCtx = &vars.cache.StatementContext // allocate from the cached data

vars.StmtCtx.RuntimeStatsColl = &vars.StmtCtx.cache.RuntimeStatsColl  // also alloc from cached data, no extra object allocation!
```

It waste a pointer size MEM, but I think that's worthy.

File: bindinfo/bind_test.go [hunk 2]
```diff
@@ -1853,8 +1853,8 @@ func (s *testSuite) TestIssue19836(c *C) {
 		PS: []*util.ProcessInfo{tk.Se.ShowProcess()},
 	})
 	explainResult := testkit.Rows(
-		"Limit_8 2.00 0 root  time:0s, loops:0 offset:1, count:2 N/A N/A",
-		"└─TableReader_14 3.00 0 root  time:0s, loops:0 data:Limit_13 N/A N/A",
+		"Limit_8 2.00 0 root   offset:1, count:2 N/A N/A",
```
Comment (tiancaiamao) [inline bindinfo/bind_test.go:1856]: Explain for connection, the RuntimeStatsColl is overwritten, so the `time:0s, loops:0` is gone.

```
select ... from ...
explain for connection ... // ResetContextOfStmt overwrite the previous statement context, so there is no runtime stats coll
```

File: executor/executor.go (hunk 1 above)
Comment (tiancaiamao) [reply executor/executor.go:1676]: > One thing I noticed is that, &xxx{} will init all fields into the default zero value.

No, all the fields are initialized to default zero value, which is done in InitStatementContext()
https://github.com/pingcap/tidb/pull/26007/files#diff-db5df2c0ea43fa82bdc165fed66584b4c5ba481fb38cf7862889fc8afa64f081R871

@xhebox

File: executor/explainfor_test.go [hunk 3]
```diff
@@ -117,8 +117,8 @@ func (s *testSerialSuite) TestExplainFor(c *C) {
 			}
 		}
 		c.Assert(buf.String(), Matches, ""+
-			"TableReader_5 10000.00 0 root  time:.*, loops:1, cop_task: {num:.*, max:.*, proc_keys: 0, rpc_num: 1, rpc_time:.*} data:TableFullScan_4 N/A N/A\n"+
-			"└─TableFullScan_4 10000.00 0 cop.* table:t1 tikv_task:{time:.*, loops:0} keep order:false, stats:pseudo N/A N/A")
+			"TableReader_5 10000.00 0 root   data:TableFullScan_4 N/A N/A\n"+
```
Comment (tiancaiamao) [inline executor/explainfor_test.go:120]: explain for connection ... the runtime stats coll is overwritten

File: executor/executor.go (hunk 1 above)
Comment (xhebox) [reply executor/executor.go:1676]: I think the following code do the same thing without adding a new field:

```
savedRuntimeStatsColl := vars.StmtCtx.RuntimeStatsColl
*vars.StmtCtx = stmtctx.StatementContext{}
vars.StmtCtx.RuntimeStatsColl = ResetRuntimeStatsColl(savedRuntimeStatsColl)
```

And I wonder the following code may have a chance not to allocate `savedSC` on heap:

```
savedSC := *vars.StmtCtx
.... // reset things and use the old SC
```

File: bindinfo/bind_test.go (hunk 2 above)
Comment (xhebox) [reply bindinfo/bind_test.go:1856]: That does not explain why it is changed. `ResetContextOfStmt` should overwrite the previous statement context, before this PR.

File: executor/builder.go [hunk 4]
```diff
@@ -970,7 +969,7 @@ func (b *executorBuilder) buildExplain(v *plannercore.Explain) Executor {
 	}
 	if v.Analyze {
 		if b.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl == nil {
-			b.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl = execdetails.NewRuntimeStatsColl()
+			b.ctx.GetSessionVars().StmtCtx.InitRuntimeStatsColl()
 		}
```
Comment (xhebox) [inline executor/builder.go:973]: Maybe changes in tests come from here, `RuntimeStatsColl` won't be nil after this PR. And it is not re-inited if `!globalConfig.EnableCollectExecutionInfo == true`. @tiancaiamao

File: executor/executor.go (hunk 1 above)
Comment (tiancaiamao) [reply executor/executor.go:1677]: Yes, you're right. @xhebox

File: executor/executor.go [hunk 5]
```diff
@@ -1793,26 +1808,30 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
...
-	if vars.StmtCtx.InUpdateStmt || vars.StmtCtx.InDeleteStmt || vars.StmtCtx.InInsertStmt {
-		sc.PrevAffectedRows = int64(vars.StmtCtx.AffectedRows())
-	} else if vars.StmtCtx.InSelectStmt {
-		sc.PrevAffectedRows = -1
-	}
 	if globalConfig.EnableCollectExecutionInfo {
-		sc.RuntimeStatsColl = execdetails.NewRuntimeStatsColl()
+		// In ExplainFor case, RuntimeStatsColl should not be reset for reuse,
+		// because ExplainFor need to display the last statement information.
+		if _, ok := s.(*ast.ExplainForStmt); ok {
```
Comment (tiancaiamao) [inline executor/executor.go:1820]: Now I do not reset  `RuntimeStatsColl` when the statement is ExplainFor, the test passes.

File: util/execdetails/execdetails.go [hunk 6]
```diff
@@ -555,9 +555,22 @@ type RuntimeStatsColl struct {
 }
 
 // NewRuntimeStatsColl creates new executor collector.
-func NewRuntimeStatsColl() *RuntimeStatsColl {
-	return &RuntimeStatsColl{rootStats: make(map[int]*RootRuntimeStats),
-		copStats: make(map[int]*CopRuntimeStats)}
+// Reuse the object to reduce allocation when *RuntimeStatsColl is not nil.
+func NewRuntimeStatsColl(reuse *RuntimeStatsColl) *RuntimeStatsColl {
+	if reuse != nil {
+		for k := range reuse.rootStats {
```
Comment (tiancaiamao) [inline util/execdetails/execdetails.go:561]: Reuse map is cheap than create a new map object.
Go compiler optimize this cleanup code pattern to a  `clearmap()` function.

File: executor/builder.go (hunk 4 above)
Comment (tiancaiamao) [reply executor/builder.go]: Now I don't reset `RuntimeStatsColl` for the `ExplainFor` cases

**Comment by tiancaiamao** [comment]
Temporary close it.
I need to reconsider this PR after I find https://github.com/pingcap/tidb/issues/26161

**Pull request closed without merging**